include/ensmallen_bits/callbacks/progress_bar.hpp

```diff
@@ -0,0 +1,103 @@
+/**
+ * @file progress_bar.hpp
+ * @author Marcus Edel
+ *
+ * Implementation of a simple progress bar callback function.
+ *
+ * ensmallen is free software; you may redistribute it and/or modify it under
+ * the terms of the 3-clause BSD license.  You should have received a copy of
+ * the 3-clause BSD license along with ensmallen.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef ENSMALLEN_CALLBACKS_PROGRESS_BAR_HPP
+#define ENSMALLEN_CALLBACKS_PROGRESS_BAR_HPP
+
+namespace ens {
+
+/**
+ * A simple progress bar, based on the maximum number of optimizer iterations
+ * and the EndEpoch callback function.
+ */
+class ProgressBar
+{
+ public:
+  /**
+   * Set up the progress bar callback class with the given width and output
+   * stream.
+   *
+   * @param width Width of the bar.
+   * @param ostream Ostream which receives output from this object.
+   */
+  ProgressBar(const size_t width = 70,
+              std::ostream& output = arma::get_cout_stream()) :
```

Comment:
`arma::get_cout_stream()` is available from Armadillo 8.100 onwards.  Currently ensmallen requires Armadillo 6.500 or later. I suggest bumping the minimum required version to 8.400, as that's available in Ubuntu 18.04 (current LTS).

Reply:
Okay, so I guess we missed this here: https://github.com/mlpack/ensmallen/blob/8bea8d214b40be3cb42e817328c0791541fbcd6c/include/ensmallen_bits/log.hpp#L74

I don't mind to bump the version here, what I could also do is to backport the functionality since it's just simple enough. Personally I don't have a strong opinion here, we would have to bump the version for mlpack as well @rcurtin do you have any preference?

Reply:
Ahh, unfortunately the reason we are stuck with 6.500 is Ubuntu 16.04 (even RHEL6 has 7.300, although I'm less concerned about RHEL6).  16.04 is still in pretty wide usage---the more recent LTS version, 18.04, has "only" been out for a year, which in my experience is not enough time for awful slow bureaucratic IT departments to upgrade.  If it's easy to backport for us, then I'd suggest that because it will make it easier for our users (who likely already have Armadillo installed or, if they are in one of these awful IT situations, they can ask the sysadmin to install it via the package manager).

In this case, I wonder if we'd be better off just defining an inline function whose implementation depends on the version of Armadillo available... if 8.100 or newer is available, it returns `arma::get_cout_stream()`, otherwise it returns `std::cout`.  What do you think?

Reply:
`arma::get_cout_stream()` has been used in ensmallen since 1.10, and so far nobody has complained :)  Maybe this indicates that Ubuntu 18.04 has spread sufficiently.

Using `std::cout` directly is problematic for the folks in the R world (ie. RcppEnsmallen and RcppArmadillo packages @coatless @eddelbuettel).  `arma::get_cout_stream()` has a configurable mechanism used by Rcpp -- allows the use of an R specific output stream.  

Otherwise any 3rd party R packages relying on `std::cout` tend to be rejected by the CRAN gatekeepers. This in turn can reduce the impact of ensmallen.

Reply:
Oh, good point, you're right and I forgot about that.  This is in `log.hpp`:

```
#ifdef ENS_PRINT_INFO
  static std::ostream& Info = arma::get_cout_stream();
#else
  static NullOutStream Info;
#endif

#ifdef ENS_PRINT_WARN
  static std::ostream& Warn = arma::get_cerr_stream();
#else
  static NullOutStream Warn;
#endif
```

So, I think already we are in a situation where if `ENS_PRINT_INFO` or `ENS_PRINT_WARN` are defined, Armadillo 8.100 or newer is required.  I looked briefly at backporting Armadillo's `get_cout_stream()` and I think it's easy*---it won't require anything other than the following code dumped into the `arma` namespace, only when the version is below 8.100:

```
template<typename T>
inline
std::ostream&
arma_cout_stream(std::ostream* user_stream)
  {
  static std::ostream* cout_stream = &(ARMA_COUT_STREAM);

  if(user_stream != NULL)  { cout_stream = user_stream; }

  return (*cout_stream);
  }



template<typename T>
inline
std::ostream&
arma_cerr_stream(std::ostream* user_stream)
  {
  static std::ostream* cerr_stream = &(ARMA_CERR_STREAM);

  if(user_stream != NULL)  { cerr_stream = user_stream; }

  return (*cerr_stream);
  }



inline
void
set_cout_stream(std::ostream& user_stream)
  {
  arma_cout_stream<char>(&user_stream);
  }



inline
void
set_cerr_stream(std::ostream& user_stream)
  {
  arma_cerr_stream<char>(&user_stream);
  }



inline
std::ostream&
get_cout_stream()
  {
  return arma_cout_stream<char>(NULL);
  }



inline
std::ostream&
get_cerr_stream()
  {
  return arma_cerr_stream<char>(NULL);
  }
```

So, perhaps that's the best way to go.  I don't have a problem with the backporting, but also we're backed into a little bit of a corner here by this situation.  So I also don't have a huge problem with bumping the minimum version to 8.100.0.  @zoq up to you.

* ignoring very strange situations that *could* happen if the user has a very weird setup.

Reply:
In general I'm not a big fan of backporting, as that creates franken-software that nobody has tested properly. In this particular case the changes are relatively small so it might be doable, though I'd still prefer bumping the minimum required version.

Reply:
I'm with @conradsnicta regarding backporting. I think it would be better to bump the required version since there are a lot of nice things in 8.yyy line (needless to say, also a licensing change). 

Also, as @conradsnicta mentioned regarding CRAN, any `std::cout` or `std::cerr` in the code will trigger an auto rejection. Details can be found at the ol' logging post: https://github.com/mlpack/ensmallen/issues/3#issuecomment-443578471

Reply:
Thanks for the input, will bump the version.